### PR TITLE
Ronald Libago - CS10 (Csaba Balogh)

### PR DESCRIPTION
**Pull request by Mister-Corn**
**This is the PR for today's Sprint Challenge, 4/20/2018**

In acknowledgment of the "3-hour time block" of these challenges, I have a timestamp on commit updates with this format "T+ (*time elapsed since start*)". Any updates after "T+ 3hr" would be labelled as "Overtime".

## Main Goals

- [x] Study Questions
- [x] Challenge 1: Objects and Arrays
- [x] Challenge 2: Functions
- [x] Challenge 3: Prototypes
- [x] Challenge 4: Classes

## Stretch Goal
- [x] A companion cube lovingly crafted with classes.
------
Student: Ronald Libago
Class: CS10
PM: Csaba Balogh

### CODE REVIEW

File: challenges/objects-arrays.js [hunk 1]
```diff
@@ -6,12 +6,22 @@
...
-// tyrannosaurus, carnivorous, 7000kg, 12m, Late Cretaceious
+class Dinosaur {
+  constructor(name, diet, weight, length, period) {
+    this.name = name;
+    this.diet = diet;
+    this.weight = weight;
+    this.length = length;
+    this.period = period;
+  }
+}
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍 I like this efficient solution.

File: challenges/objects-arrays.js [hunk 2]
```diff
@@ -6,12 +6,22 @@
...
 
+// tyrannosaurus, carnivorous, 7000kg, 12m, Late Cretaceious
+const tyrannosaurus = new Dinosaur(`Tyrannosaurus`, `carnivorous`, `7000kg`, `12m`, `Late Cretaceious`)
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍 A good way to code classes if you have a repetitive assignment.

File: challenges/objects-arrays.js [hunk 3]
```diff
@@ -6,12 +6,22 @@
...
 // stegosaurus, herbivorous, 2000kg, 9m, Late Jurassic
-
+const stegosaurus = new Dinosaur(`Stegosaurus`, `herbivorous`, `2000kg`, `9m`, `Late Jurassic`)
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 4]
```diff
@@ -6,12 +6,22 @@
...
 // velociraptor, carnivorous, 15kg, 1.8m, Late Cretaceious
-
+const velociraptor = new Dinosaur(`Velociraptor`, `carnivorous`, `15kg`, `1.8m`, `Late Cretaceious`)
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 5]
```diff
@@ -24,6 +34,10 @@ console.log(stegosaurus.length);
 console.log(tyrannosaurus.period);
 
 // Create a new roar method for the tyrannosaurus.  When called, return "RAWERSRARARWERSARARARRRR!" Log the result.
+tyrannosaurus.roar = function() {
+  return "RAWERSRARARWERSARARARRRR!";
+}
+
 console.log(tyrannosaurus.roar());
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 6]
```diff
@@ -46,6 +60,14 @@ const graduates = [{"id":1,"first_name":"Cynde","university":"Missouri Southern
...
 const universities = [];
+
+function each(baseArray, baseProperty, newArray) {
+  for (let obj of baseArray) {
+    newArray.push(obj[baseProperty]);
+  }
+  return newArray;
+}
+each(graduates, "university", universities);
 console.log(universities)
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: Let's not forgot to sort it alphabetically

File: challenges/objects-arrays.js [hunk 7]
```diff
@@ -54,12 +76,35 @@ The resulting contact information should have a space between the first name and
 Name email@example.com
 
 Log the result of your new array. */
+function dobleEach(baseArray, basePropertyA, basePropertyB, newArray) {
+  for (let obj of baseArray) {
+    newArray.push(`${obj[basePropertyA]} ${obj[basePropertyB]}`);
+  }
+  return newArray;
+}
+
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍 
Works well.

File: challenges/objects-arrays.js [hunk 8]
```diff
@@ -54,12 +76,35 @@ The resulting contact information should have a space between the first name and
...
 
 
 /* Request 3: Find out how many universities have the string "Uni" included in their name. Create a new array called uni that contains them all. Log the result. */
-const uni = [];
+function jankFilter(array, callback) {
+  let holdArray = []
+  for (let element of array) {
+    if (callback(element)) {
+      holdArray.push(element);
+    } else {
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: this part is redundant, you can skip else { continue }, the result will be the same.

Comment (JasonCorp84) [reply challenges/objects-arrays.js]: We have all the universities now that contains "Uni", but we have to count them.
Let's also log how many we have.
👍

File: challenges/objects-arrays.js [hunk 9]
```diff
@@ -79,9 +124,9 @@ zooAnimals = [{"animal_name":"Jackal, asiatic","population":5,"scientific_name":
 {"animal_name":"Australian pelican","population":5,"scientific_name":"Pelecanus conspicillatus","state":"West Virginia"}];
 
 // The zoos need a list of all their animal's names converted to lower case.  Create a new array named lowerCase and map over each name to convert them all to lower case.  Log the resut.
-let lowerCase = [];
+let lowerCase = zooAnimals.map(animal => animal.animal_name.toLowerCase());
 console.log(lowerCase);
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/objects-arrays.js [hunk 10]
```diff
@@ -79,9 +124,9 @@ zooAnimals = [{"animal_name":"Jackal, asiatic","population":5,"scientific_name":
 {"animal_name":"Australian pelican","population":5,"scientific_name":"Pelecanus conspicillatus","state":"West Virginia"}];
 
 // The zoos need a list of all their animal's names converted to lower case.  Create a new array named lowerCase and map over each name to convert them all to lower case.  Log the resut.
-let lowerCase = [];
+let lowerCase = zooAnimals.map(animal => animal.animal_name.toLowerCase());
 console.log(lowerCase); 
 
 // The zoos need to know their total animal population across the United States.  Add up all the population numbers from all the zoos using the .reduce() method.
-let populationTotal = [];
+let populationTotal = zooAnimals.reduce((acc, animal) => acc + animal.population, 0);
```
Comment (JasonCorp84) [inline challenges/objects-arrays.js]: 👍

File: challenges/prototypes.js [hunk 11]
```diff
@@ -5,29 +5,41 @@
 /* == Step 1: Base Constructor ==
   Create a constructor function named CuboidMaker that accepts properties for length, width, and height
 */
-
+function Cuboid(length, width, height) {
+  this.length = length;
+  this.width = width;
+  this.height = height;
+}
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: 👍

File: challenges/prototypes.js [hunk 12]
```diff
@@ -5,29 +5,41 @@
...
 
 /* == Step 2: Volume Method ==
   Create a method using CuboidMaker's prototype that returns the volume of a given cuboid's length, width, and height
   
   Formula for cuboid volume: length * width * height
 */
 
+Cuboid.prototype.volume = function() {
+  return this.length * this.width * this.height;
+}
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: 👍

File: challenges/prototypes.js [hunk 13]
```diff
@@ -5,29 +5,41 @@
...
 
 /* == Step 3: Surface Area Method ==
   Create another method using CuboidMaker's prototype that returns the surface area of a given cuboid's length, width, and height. 
 
   Formula for cuboid surface area of a cube: 2 * (length * width + length * height + width * height)
 */
 
+Cuboid.prototype.surfaceArea = function() {
+  return 2 * (this.length * this.width + this.length * this.height + this.width * this.height);
+}
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: 👍

File: challenges/prototypes.js [hunk 14]
```diff
@@ -5,29 +5,41 @@
...
 
 /* == Step 4: Create a new object that uses CuboidMaker ==
-  Create a cuboid object that uses the new keyword to use our CuboidMaker constructor
+  Create a COMPANION CUBE object that uses the new keyword to use our CuboidMaker constructor
   Add properties and values of length: 4, width: 5, and height: 5 to cuboid.   
 */
 
+const companionCube = new Cuboid(4, 5, 5);
+
```
Comment (JasonCorp84) [inline challenges/prototypes.js]: The cake is a lie. 👍

File: challenges/functions.js [hunk 15]
```diff
@@ -6,27 +6,49 @@
...
+console.log(`\n ------------ Step 1 ------------ \n`);
 
+function consume(a, b, callback) {
+  return callback(a, b);
+}
+
+const aStr = "So it's not just gonna happen like that";
+const bStr = `'Cause I ain't no`;
+const hollaBack = (a, b) => `${a}, ${b} hollaback guuuurrrl`;
+console.log(consume(aStr, bStr, hollaBack));
```
Comment (JasonCorp84) [inline challenges/functions.js]: 👍

File: challenges/functions.js [hunk 16]
```diff
@@ -6,27 +6,49 @@
...
 /* Step 2: Create several functions to callback with consume();
   * Create a function named add that returns the sum of two numbers
   * Create a function named multiply that returns the product of two numbers 
   * Create a function named greeting that accepts a first and last name and returns "Hello first-name last-name, nice to meet you!"
 */
+console.log(`\n ------------ Step 2 ------------ \n`);
 
+const add = (a, b) => a + b;
+const multiply = (a, b) => a * b;
+const greeting = (firstName, lastName) => `Hello ${firstName} ${lastName}, nice to meet you!`;
```
Comment (JasonCorp84) [inline challenges/functions.js]: 👍 I like this code. Easy to read, just like a simple sentence.

File: challenges/functions.js [hunk 17]
```diff
@@ -6,27 +6,49 @@
...
-// consume(10,16,multiply); // 160
-// consume("Mary","Poppins", greeting); // Hello Mary Poppins, nice to meet you!
+console.log(`\n ------------ Step 3 ------------ \n`);
 
+consume(2,2,add); //? 
+// 4 
+consume(10,16,multiply); //?
+// 160
+consume("Mary","Poppins", greeting); //?
+// Hello Mary Poppins, nice to meet you!
```
Comment (JasonCorp84) [inline challenges/functions.js]: 👍

File: challenges/functions.js [hunk 18]
```diff
@@ -6,27 +6,49 @@
...
+
+// However, JavaScript tries *real* hard to be a bro, and so instead of just giving up there and then, the engine tries to find the value in the variable environment that the function lexically is in. That's a fancy way of saying look at the variable environment of whatever the function is physically sitting in the code. In this case, `nestedFunction()` sits within `myFunction()`, so we can look at the variable environment of `myFunction()`. Fortunately, we find the value of `internal` there, which is `Hello! I'm inside the function`. Therefore, `nestedFuction()` outputs `Hello! I'm inside the function`.
+
+// If it didn't find it there, the engine will continue to look outside to the variable environment of the context `myFunction()` lexically sits in. It will keep doing this until it hits global. If it still can't find it, it will finally give up and throw a `Reference Error`.
```
Comment (JasonCorp84) [inline challenges/functions.js]: Great explanation. I like the fact that you explain what JavaScript engine does. Way to go!

File: challenges/classes.js [hunk 19]
```diff
@@ -1,8 +1,65 @@
...
+    this.width = width;
+    this.height = height;
+  }
+  volume() {
+    return this.length * this.width * this.height;
+  }
+  surfaceArea() {
+    return 2 * (this.length * this.width + this.length * this.height + this.width * this.height);
+  }
+}
```
Comment (JasonCorp84) [inline challenges/classes.js]: 👍

File: Answers.md [hunk 20]
```diff
@@ -1,6 +1,72 @@
 # Your responses to the short answer questions should be laid out here using Mark Down.
 1. Describe the biggest difference between `.forEach` & `.map`.
+
+`.forEach` simply performs the callback result on each element of the array.
+* Because of this, `.forEach` is great for doing side effects, like `console.log()`, or pushing/popping things from arrays.
+* Because it doesn't return anything, some kind of mutation is necessary to see its results. If you want to iterate over an array and do the callback function on each element, you'd have to change the array you're iterating over, or initialize a blank new array and change that to see any output.
+
+`.map` takes the each result of the callback functions and returns all of the results into a new array.
+* Because of this, this doesn't need to change the array it is working on, nor do you need to set up an output array of your own. `map()` automatically returns a new array filled with results of the callback function iterating over the original array.
+
```
Comment (JasonCorp84) [inline Answers.md]: 👍 clear explanation!

File: Answers.md [hunk 21]
```diff
@@ -1,6 +1,72 @@
...
+Method:
+```js
+const foo = {
+    bar: function () {
+        return 'Hello';
+    },
+}
+foo.bar();
+```
+
```
Comment (JasonCorp84) [inline Answers.md]: Always great to have examples: keep it up!

File: Answers.md [hunk 22]
```diff
@@ -1,6 +1,72 @@
...
+}
+
+let whatAreTheyDoing = room();
+console.log(whatAreTheyDoing());
+```
+
+In the body of the function definition, we need the values of `personA` and `personB`. But, we only have the value of `personB` available right in that function. Closure allows us to go outside of the function and to the variable environment of the 'parent', which is function `room()`. It does have the value for `personA`, so the engine will proceed to evaluate `personA + personB` => `34 + 35` and return `69`. If it couldn't find `personA` in `room()`, then the engine would continue to look outward until the hits the global scope. If it fails in its search, only then will it throw a `Reference Error`.
+
+This particular snippet points to something a little more esoteric than the above explanation, but I think this works nicely.
+
```
Comment (JasonCorp84) [inline Answers.md]: 👍 👍 👍

File: Answers.md [hunk 23]
```diff
@@ -1,6 +1,72 @@
...
+If the context you're working on is the global context, `this` will point to the global object. This is kinda why this:
+* Implicit Binding
+If you call a method, `this` points to the object the method is assocated with.
+
+You can direct what `this` points to. This covers the below principles:
+* Explicit Binding
+You can tell JavaScript what `this` should refer to with `call()`, `apply()`, and `bind()`.
+* `new` Keyword Binding
+When you use the `new` keyword when you use a constructor function, `new` redirects `this` to the new instance of the object, instead of letting it stay directed to the object of the constructor function (*which is probably global*).
+
```
Comment (JasonCorp84) [inline Answers.md]: 👍

File: Answers.md [hunk 24]
```diff
@@ -1,6 +1,72 @@
...
+5. Why do we need super() in an extended class?
+
+`super()` is the newer way of saying `ParentClass.call(this, attributes);`. In other words, `super()` calls the base class' constructor function, binds `this` to the extended Class, then uses that constructor function to populate the new object with properties from the base class, assigning values passed as arguments into `super()` in the extended class.
```
Comment (JasonCorp84) [inline Answers.md]: 👍 👍 👍

**Comment by mixelpixel** [comment]
@Mister-Corn AWESOME time keeping msgs!!!

It's great to see that you are committing frequently. You get a gold ⭐️ !! Glad to see you are already so comfortable with the Git and GitHub tools :octocat: 🎉 

Looking ahead to the larger projects you'll be working on, I'd ❤️ to see your msgs conveying more information, more frequently. This habit will make you a better programmer! Any time you complete an assignment, a function, add a component, or try a new feature in your programs, if you make note of the updates by using Git commit messages, your Git history will _**tell the story of your software engineering adventure!**_

While the Git commit titles are short, you can add a "body" to the messages which can be much more detailed. BEFORE you add the closing quote to the commit message, hit enter twice to add a new line between the title and the body. For example:

```
$ git commit -m "This is a Git commit TITLE less than 72 characters

This is a Git commit BODY which can contain as much pertinent information and descriptions as you see fit. Some development teams like tons and tons of information, and depending on the scope of the commit, it may be necessary to make lots of detailed notes - especially in projects like React where there can be a vast amount of files and a complex project architecture.
- I did this.
- I did that.
- this happens.
- that needs work.
- this thing got refactored.
- where's my Lambda swag?????"
```

...on GitHub in the Pull Requests, the body will be accessible by clicking on the three little dots:

triple-dots

...such that they will reveal the BODY:

reveal

If you're not already in the #git channel in Slack, come check it out: https://lambdaschoolstudents.slack.com/archives/C5Z8BMX6K/p1523565197000063

😺

**Pull request closed without merging**